### ISSUE TITLE
unexpected result with groupby apply on categorical data

### ISSUE THREAD

**Issue by tmoroz** [pydata/pandas#9603]
```python
df = pandas.DataFrame({'a': [1, 0, 0, 0]})
df.groupby(pandas.cut(df.a, [0, 1, 2, 3, 4])).apply(lambda x: len(x))

Out[3]:
a
(0, 1]    1
(1, 2]    0
(2, 3]    0
(3, 4]    3
dtype: int64
```
final group length should be 0 not 3

**Comment by jreback** [comment]
I recall another issue that is similar (but can't find ATM).

You can use these (and are much faster anyhow)
```
In [6]: df.groupby(pandas.cut(df.a, [0, 1, 2, 3, 4])).size()
Out[6]: 
a
(0, 1]     1
(1, 2]   NaN
(2, 3]   NaN
(3, 4]   NaN
dtype: float64

In [7]: df.groupby(pandas.cut(df.a, [0, 1, 2, 3, 4])).count()
Out[7]: 
        a
a        
(0, 1]  1
(1, 2]  0
(2, 3]  0
(3, 4]  0
```

**Comment by jreback** [comment]
closed by #10014
